include/jemalloc/internal/cache_bin.h

```diff
@@ -441,29 +441,18 @@ cache_bin_finish_fill(cache_bin_t *bin, cache_bin_info_t *info,
 	bin->stack_head = empty_position - nfilled;
 }
 
-/* Same deal, but with flush. */
+/*
+ * Same deal, but with flush.  Unlike fill (which can fail), the user must flush
+ * everything we give them.
+ */
 static inline void
 cache_bin_init_ptr_array_for_flush(cache_bin_t *bin, cache_bin_info_t *info,
     cache_bin_ptr_array_t *arr, cache_bin_sz_t nflush) {
-	arr->ptr = cache_bin_empty_position_get(bin) - 1;
+	arr->ptr = cache_bin_empty_position_get(bin) - nflush;
 	assert(cache_bin_ncached_get_local(bin, info) == 0
 	    || *arr->ptr != NULL);
 }
 
-/*
- * These accessors are used by the flush pathways -- they reverse ordinary array
```

Comment:
The change on the reserver order made this part better in every way. Which made me wonder, was there any particular reason this was reverse flushed? Or was it really just out of "convenience"?

Reply:
I think it's mostly a historical accident. Originally, there was just a for loop in the arena code accessing tcache internals directly, where it was just as easy to write `for (void **ptr = bin_end; ptr != bin_end - nflush; ptr--)`.

When extracting cache_bin out of the tcache/arena interactions, I accidentally inverted the flush ordering during development and it caused a hard-to-track-down regression. So after that, I just kept every access the exact same, even at the cost of extra syntactic overhead. I don't think anyone ever consciously decided that the syntactic overhead was necessary to accomplish some other goal (except for accommodating my laziness).